nucypher/cli/characters/alice.py

```diff
@@ -11,171 +12,411 @@
+                                                   registry_filepath=registry_filepath,
+                                                   provider_process=ETH_NODE,
+                                                   poa=poa,
+                                                   provider_uri=provider_uri,
+                                                   m=m,
+                                                   n=n,
+                                                   duration_periods=duration_periods,
+                                                   rate=rate)
+
+    painting.paint_new_installation_help(emitter, new_configuration=new_alice_config)
+
+
+@alice.command()
+@click.option('--config-file', help="Path to configuration file", type=EXISTING_READABLE_FILE)
+@nucypher_click_config
+def view(click_config, config_file):
+    """
+    View existing Alice's configuration.
+    """
+    emitter = _setup_emitter(click_config)
+
+    configuration_file_location = config_file or AliceConfiguration.default_filepath()
+    response = AliceConfiguration._read_configuration_file(filepath=configuration_file_location)
+    return emitter.ipc(response=response, request_id=0, duration=0)  # FIXME: what are request_id and duration here?
+
 
+@alice.command()
+@_admin_options
+@click.option('--config-file', help="Path to configuration file", type=EXISTING_READABLE_FILE)
+@click.option('--discovery-port', help="The host port to run node discovery services on", type=NETWORK_PORT)
+@click.option('--force', help="Don't ask for confirmation", is_flag=True)
+@nucypher_click_config
+def destroy(click_config,
+
+            # Admin Options
+            geth, provider_uri, federated_only, dev, pay_with, network, registry_filepath,
+
+            # Other
+            config_file, discovery_port, force):
+    """
+    Delete existing Alice's configuration.
+    """
+    ### Setup ###
+    emitter = _setup_emitter(click_config)
+
+    alice_config, provider_uri = _get_alice_config(click_config, config_file, dev, discovery_port, federated_only,
+                                                   geth, network, pay_with, provider_uri, registry_filepath)
+    #############
+
+    if dev:
+        message = "'nucypher alice destroy' cannot be used in --dev mode"
+        raise click.BadOptionUsage(option_name='--dev', message=message)
+    return actions.destroy_configuration(emitter, character_config=alice_config, force=force)
+
+
+@alice.command()
+@_api_options
+@click.option('--controller-port', help="The host port to run Alice HTTP services on", type=NETWORK_PORT,
+              default=AliceConfiguration.DEFAULT_CONTROLLER_PORT)
+@click.option('--dry-run', '-x', help="Execute normally without actually starting the node", is_flag=True)
+@nucypher_click_config
+def run(click_config,
+
+        # API Options
+        geth, provider_uri, federated_only, dev, pay_with, network, registry_filepath,
```

Comment:
These repeating argument lists kind of bother me. Do you think it would be better to gather them into structures e.g.

```
def _admin_options2(func):
    @click.option('--geth', '-G', help="Run using the built-in geth node", is_flag=True)
    @click.option('--provider', 'provider_uri', help="Blockchain provider's URI", type=click.STRING)
    @click.option('--federated-only', '-F', help="Connect only to federated nodes", is_flag=True)
    @click.option('--dev', '-d', help="Enable development mode", is_flag=True)
    @click.option('--pay-with', help="Run with a specified account", type=EIP55_CHECKSUM_ADDRESS)
    @click.option('--network', help="Network Domain Name", type=click.STRING)
    @click.option('--registry-filepath', help="Custom contract registry filepath", type=EXISTING_READABLE_FILE)
    @functools.wraps(func)
    def wrapper(
            geth,
            provider_uri,
            federated_only,
            dev,
            pay_with,
            network,
            registry_filepath,
            **kwargs):

        opts = namedtuple(
            'AdminOptions',
            'geth',
            'provider_uri',
            'federated_only',
            'dev',
            'pay_with',
            'network',
            'registry_filepath',
            )

        return func(admin_options=opts, **kwargs)

    return wrapper


def _api_options2(func):
    @_admin_options2
    @click.option('--config-file', help="Path to configuration file", type=EXISTING_READABLE_FILE)
    @click.option('--discovery-port', help="The host port to run node discovery services on", type=NETWORK_PORT)
    @click.option('--hw-wallet/--no-hw-wallet', default=False)
    @click.option('--teacher', 'teacher_uri', help="An Ursula URI to start learning from (seednode)", type=click.STRING)
    @click.option('--min-stake', help="The minimum stake the teacher must have to be a teacher", type=click.INT,
                  default=0)
    @functools.wraps(func)
    def wrapper(
            admin_options,
            config_file,
            discovery_port,
            hw_wallet,
            teacher,
            min_stake,
            **kwargs):

        opts = namedtuple(
            'ApiOptions',
            'admin_options',
            'discovery_port',
            'hw_wallet',
            'teacher',
            'min_stake',
            )

        return func(api_options=opts, **kwargs)

    return wrapper
```

and then

```
@alice.command()
@_admin_options2
@click.option('--config-root', help="Custom configuration directory", type=click.Path())
@click.option('--poa', help="Inject POA middleware", is_flag=True, default=None)
@click.option('--m', help="M-Threshold KFrags", type=click.INT)
@click.option('--n', help="N-Total KFrags", type=click.INT)
@click.option('--rate', help="Policy rate per period in wei", type=click.FLOAT)
@click.option('--duration-periods', help="Policy duration in periods", type=click.FLOAT)
@nucypher_click_config
def init(click_config,
         admin_options,
         # Other
         config_root, poa, m, n, rate, duration_periods):
```

Or is it too much decorator magic?

Reply:
👍  - It would be cleaner and prevent copy and pasting - Filed #1431 .